etcd/etcd.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2023 CloudWeGo Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcd
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+	"text/template"
+
+	"github.com/cloudwego/kitex/pkg/klog"
+	"go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/mvcc/mvccpb"
+	"go.uber.org/zap"
+)
+
+var (
+	m      sync.Mutex
+	ctxMap map[string]context.CancelFunc
+)
+
+type Key struct {
+	Prefix string
+	Path   string
+}
+
+type Client interface {
+	SetParser(ConfigParser)
+	ClientConfigParam(cpc *ConfigParamConfig, cfs ...CustomFunction) (Key, error)
+	ServerConfigParam(cpc *ConfigParamConfig, cfs ...CustomFunction) (Key, error)
+	RegisterConfigCallback(ctx context.Context, cancel context.CancelFunc, key string, callback func(string, ConfigParser))
+	DeregisterConfig(key string)
+}
+
+type client struct {
+	ecli *clientv3.Client
+	// support customise parser
+	parser             ConfigParser
+	prefixTemplate     *template.Template
+	serverPathTemplate *template.Template
+	clientPathTemplate *template.Template
+}
+
+// Options etcd config options. All the fields have default value.
+type Options struct {
+	Address          string
+	Port             uint64
+	Prefix           string
+	ServerPathFormat string
+	clientPathFormat string
+	LoggerConfig     *zap.Config
+	ConfigParser     ConfigParser
+}
+
+// New Create a default etcd client
+// It can create a client with default config by env variable.
+// See: env.go
+func New(opts Options) (Client, error) {
+	if opts.Address == "" {
+		opts.Address = EtcdDefaultServerAddr
+	}
+	if opts.Port == 0 {
+		opts.Port = EtcdDefaultPort
+	}
+	if opts.ConfigParser == nil {
+		opts.ConfigParser = defaultConfigParse()
+	}
+	if opts.Prefix == "" {
+		opts.Prefix = EtcdDefaultConfigPrefix
+	}
+	if opts.ServerPathFormat == "" {
+		opts.ServerPathFormat = EtcdDefaultServerPath
+	}
+	if opts.clientPathFormat == "" {
+		opts.clientPathFormat = EtcdDefaultClientPath
+	}
+	etcdClient, err := clientv3.New(clientv3.Config{
+		Endpoints: []string{fmt.Sprintf("http://%s:%d", opts.Address, opts.Port)},
```

Comment:
`Endpoints` could also be an array of the etcd instances, so we should provide a way for the user to specify multiple endpoints